examples/demo/piano.html

```diff
@@ -0,0 +1,219 @@
+      length: 12 * nKeys,
+      expr: function (emit, x, t) {
+        r = 2/(nKeys*nRepeats * (1 + keyGap )) ;  // r is the unit of size.
+        p = Math.floor( x );  
+
+        face = p % 6;           // Face number 0..5 for this cuboid.
+        p = Math.floor( p/6);   // Cuboid number.
+        
+        yDisplace = 0;
+        keyWidth  = 1.0;
+        keyDepth  = keyDepth1;
+        xDisplace = -1/(r*nRepeats); // Start at left.      
+
+        // 2 cuboids per key.
+        key = Math.floor( p/2);      // Key number.
+
+        xDisplace += key*(1+keyGap); // Keys are 1 unit apart, plus a gap.
+        key = key %7;                // Key shapes repeat after 7.
+        
+        // Odd numbered cuboids are amongst the black keys
+        // and are longer, further back, displaced and narrower
+        // than the cuboid at the 'front' of the key.
+        if( (p%2) == 1 ){
+          keyDepth = keyDepth2;  // longer.
+          yDisplace = keyDepth1; // Further back.
+          
+          // All but 0 & 4 are displaced half a black key width to the right.
+          if( (key % 4) != 0 )
+            xDisplace += 0.25+keyGap/2;
+            
+          // 0,3,4 and 6 are slightly wider.  1,2 & 5 narrow.
+          if( (key==0) || (key ==3) || (key==4) || (key==6) )
+            keyWidth = 0.75-keyGap/2;
+          else {
+            keyWidth = 0.5-keyGap;
+          }
+        }
+        // r is the unit of size.
+        cuboid_face( emit, -r*keyWidth, -r*keyDepth, r/4, -r*xDisplace, -r*yDisplace, 0, face );
+      },
+      items: 4,    // because we output four points for a face.
+      channels: 3, // because we output xyz.
+    });
+   
+    // We are using the items dimension, 4 items, to hold 4 points for a face.
+   
+    // 'repeat' replicates in one of the dimensions.
+    // We were only using the width and items dimensions, so this
+    // repeat on 'height' adds a new dimension to our array of points.
+    // If nRepeats were 3 we would have 3 copies of the data.
+    view.repeat({
+      height: nRepeats,
+    });
+    // 'spread' transfers information from our array dimensions into 
+    // the vec4's actually in the array.
+    // We're spreading based on the newly added dimension.
+    // That's quite a common pattern.
+    view.spread({
+      unit: 'absolute',
+      height: [2/nRepeats, 0, 0, 0], // This means we will affect the 'x' of the vec4s.
+    });
+    
+    // The earlier emits put values in to x,y,z, but the way we are viewing this,
+    // our 'y' goes into the page and our z goes up.  We wouldn't need this swizzle
+    // if we'd emitted xyz in a different order.
+    view.swizzle({
+      id : 'whiteKeyPoints',
+      order : 'yzxw',
+    });
+ 
+    // Draw the white keys.
+    // Take groups of four items, from the full width x height array of points, and
+    // Treat those groups as faces (rectangles).
+    view.face({
+      points :'#whiteKeyPoints',
```

Comment:
It's a matter of style, but adding an id when you don't need it creates some visual noise.